array.rb

```diff
@@ -1,4 +1,16 @@
 class Array
+  def each
+    unless block_given?
+      return to_enum(:each) { self.length }
+    end
+    i = 0
+    while i < self.length
+      yield self[i]
```

Comment:
This is likely not thread-safe, because the GVL could be released between `i < self.length` and `self[i]` (AFAIK), and another thread could e.g. shrink the array, and then `self[i]` is `nil` (even though that Array would contain no `nil` element), while when it was implemented in C this could not happen.

Reply:
There's probably no efficient way in Ruby on CRuby to fix it. Are you proposing to add a Primitive to support the edge case?

Reply:
In TruffleRuby this is one of the reasons Array#each is implemented in Java, because we need finer control of interrupt checks/safepoints for it (and `Thread.handle_interrupt` is horribly inefficient).
Also for the case of an Array shared between threads on TruffleRuby, we'll likely need some logic that can't be expressed in Ruby to avoid that problem (that part is not merged in truffleruby yet but is part of my PhD work).

I think it's an example of why some core methods cannot really be implemented in Ruby due to incompatibilities or not being able to express something with Ruby code.

Reply:
TruffleRuby can mix and optimize Ruby and Java together, but CRuby can't do the same thing with Ruby and C. We need this line to be `yield` (`invokeblock`) to transform it with a JIT. I guess I can put `Primitive` around the `yield` call to deal with it then.

Reply:
Right.

> I guess I can put `Primitive` around the `yield` call to deal with it then.

What do you mean?
I think you'd need to do `_len = self.length`, `i < _len` and `self[i]` all "atomically", i.e., without releasing the GIL and without interrupt checks for compatibility (I think otherwise this incompatibility could break some Ruby programs).

Reply:
The most complicated part would be to return `i < self.length` and `self[i]` from a single C function. I guess you could return one of these as a return value of the C function, and assign the other to a local variable and reference it.

Reply:
One thought is one could use `Qundef` or so to mean "no more elements" and so just the return value would be enough for both.
But also it means YJIT wouldn't see through that Primitive, unless it intrinisifies it, right?
Doesn't prevent inlining/optimizing the block call but might be counter-productive for the overall performance of Array#each.

Reply:
Interesting idea, but you're right; `Primitive`s will be regular C function calls by default, which would be slower than method calls optimized by JIT.

Another idea, in pure Ruby:

```rb
class Array
  def each
    unless block_given?
      return to_enum(:each) { self.length }
    end
    i = 0
    while (val = self[i]; i < self.length)
      yield val
    end
  end
end
```

It doesn't work if you override `[]` or `<`, but it should otherwise work.

Reply:
That's still not thread-safe of course, e.g. if between `val = self[i];` and `i < self.length` another thread appends an item to the array this will yield an extra `nil`.

Reply:
> if between val = self[i]; and i < self.length another thread appends an item to the array this will yield an extra nil.

In CRuby, those operations don't incur `VM_CHECK_INTS` in my understanding, unless `Array#[]`, `Integer#<`, or `Integer#length` is overridden. In the other thread https://github.com/ruby/ruby/pull/6687#discussion_r1093672235, we already discussed that we don't want to support redefinition of every single method here. With that assumption, it should work for CRuby.

Reply:
I'm not thinking about redefinition, that's a different discussion indeed.

Right, then it may be enough. It's a bit brittle as what checks interrupts or not may change I'd think but probably not often (would be good to add a test).
Another thought: could some TracePoint (e.g. line, call or c_call) cause the inconsistency here (since the TracePoint will perform extra interrupt checks)?
It doesn't seem nice if this bug could happen when e.g. a debugger using TracePoint is used, but would work fine without the debugger (i.e. with no TracePoint).

Reply:
With TracePoint enabled, those methods will be regular method calls https://github.com/ruby/ruby/blob/59c3fac6c4d803019095eebb92b0d2862450ded6/tool/ruby_vm/views/_trace_instruction.erb#L14-L17 so that'd be broken.

IIRC byebug stops all other threads. If debug.gem does the same thing by default, it might not practically happen. But if those assumptions are not correct, I guess we'll need a Primitive call.

Reply:
> IIRC byebug stops all other threads. If debug.gem does the same thing by default, it might not practically happen. But if those assumptions are not correct, I guess we'll need a Primitive call.

There could be other usages of TracePoint too which could then break this.
Using the example of a debugger, in Java debuggers at least it's possible to specify whether a breakpoint stops all threads or just the current thread (this is useful to debug threading issues, I recently used it).

BTW how could byebug stop all other threads? AFAIK there is no way to prevent thread/GVL switching on CRuby (and the debugger will likely wait in `IO#gets` for input which releases the GVL).